### ISSUE TITLE
julia-vscode crash after I compiled a new system image with OhMyREPL

### ISSUE THREAD

**Issue by Roger-luo** [julia-vscode/julia-vscode#1038]
So I basically followed this: https://julialang.github.io/PackageCompiler.jl/dev/examples/ohmyrepl/

then my vscode start to crash, it's version 0.14.13 at the moment.

**Comment by davidanthoff** [comment]
You mean the language server crashes?

If I understand those instructions properly, you are replacing the system image, right? So I guess that is a situation where our isolation of the extension from what users do on their system doesn't really work because we will be started with that modified sysimage... 

I think generally it might just not be a good idea to replace the sysimage, as then all Julia instances will have to deal with that new sysimage. A better option might be to use VS Code's ability to compile a sysimage for a given env, which will then be loaded whenever you open a REPL inside Julia.

**Comment by Roger-luo** [comment]
True, thanks for the explanation! I replaced the sysimg indeed... didn't expect it would crash because of that lol. But I'm wondering if it's possible to use OhMyREPL inside vscode tho? if vscode keep using its own sysimg.

**Comment by davidanthoff** [comment]
So in general there is a command to compile a sysimage for the current environment in the extension. Once you've done that, if you start a REPL, it will use that sysimage. So maybe it might be enough to add OhMyREPL to the current environment and then compile a sysimage? But probably not... @KristofferC might have some ideas :)

**Comment by KristofferC** [comment]
As long as the same versions of all dependencies are used in the sysimage as Vscode use, things should be compatible and not crash. Any idea what is causing the crash?

**Comment by davidanthoff** [comment]
> As long as the same versions of all dependencies are used in the sysimage as Vscode use

That is almost certainly not the case. The language server that runs in the VS Code extension runs with its own versions of all the Julia packages that it needs (that process is even started with a custom `JULIA_DEPOT_PATH` to make it as independent from anything that a user might have in their own `.julia` folder as possible).

If replacing the default sysimage implies that one can't load an environment anymore that has different versions of packages than what is in the sysimage, then it seems in general a bad idea to actually replace the default sysimage? In particular if there is then no way left that "we" (the VS Code extension) can avoid that custom sysimage.

I think the way to still use custom sysimages in this situation is to not replace the default one, but instead create a new one that is then loaded explicitly for specific Julia instances (which happens automatically if the sysimage was compiled with the VS Code extension).

**Comment by KristofferC** [comment]
> If replacing the default sysimage implies that one can't load an environment anymore that has different versions of packages than what is in the sysimage, then it seems in general a bad idea to actually replace the default sysimage? In particular if there is then no way left that "we" (the VS Code extension) can avoid that custom sysimage.

Yeah, there are drawbacks to replacing the default sysimage. A better idea is probably to locally define something like

```
alias julia = /usr/bin/julia -J/Users/kc/my_default_sysimage.dylib`
```

or something like that.

**Comment by davidanthoff** [comment]
I guess that would work, as long as one configured the extension to use `/usr/bin/julia` instead of `julia` to start the Julia executable...

But I think in the context of VS Code, it is much easier to use the built in command to compile a sysimage for an env because that will then be loaded automatically for any REPL instance in VS Code. That should also work for different environments (i.e. one can compile a separate sys image for different environments, and it should automatically load the right one.

It would be really nice if the normal Julia REPL used the same logic to auto-load sysimages.

**Comment by marius311** [comment]
> it is much easier to use the built in command to compile a sysimage for an env because that will then be loaded automatically for any REPL instance in VS Code

The problem with this feature for me is that I like to replace my default sysimage with one that has _just a few_ extra packages to get snappier startup, usually IJulia and Revise, both of which come from my `v1.X` global environment (and are basically the only things in that environment, all my developement happens in separate environments with all their dependencies tracked there). 

Is there any way to do something like this in a way thats compatible with this plugin?

**Comment by davidanthoff** [comment]
No, I don't think so.

**Comment by pkienscherf** [comment]
> Is there any way to do something like this in a way thats compatible with this plugin?

What seems to work is to create such a sysimage with the VSCode extension (e.g., in a new project that has just the packages you _always_ need), and then replace the default image with that one. I don't really assume that this is a particularly stable solution, but for the most part, it does what you asked.

**Comment by gabehassler** [comment]
> What seems to work is to create such a sysimage with the VSCode extension (e.g., in a new project that has just the packages you always need), and then replace the default image with that one. 
@pkienscherf Can you say how you did that? I replaced the original `sys.dll` file with the `JuliaSysimage.dll` file created by the VSCode extension, but the language server still crashes.
